File: app/local_plugins/device_jieli/android/src/main/kotlin/com/jielihome/jielihome/bridge/MethodRouter.kt
package com.jielihome.jielihome.bridge

import android.content.Context
import com.jielihome.jielihome.core.JieliHomeServer
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * MethodChannel 路由层：把 Dart 调用映射到 server 内的 feature 模块。
 * Plugin 类不直接处理任何业务逻辑，只装配 router。
 */
class MethodRouter(
    private val context: Context,
    private val server: JieliHomeServer,
) : MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "getPlatformVersion" ->
                    result.success("Android ${android.os.Build.VERSION.RELEASE}")

                "initialize" -> {
                    server.initialize(
                        context = context,
                        multiDevice = call.argument<Boolean>("multiDevice") ?: true,
                        skipNoNameDev = call.argument<Boolean>("skipNoNameDev") ?: false,
                        enableLog = call.argument<Boolean>("enableLog") ?: false,
                    )
                    result.success(true)
                }

                "startScan" -> {
                    val timeout = call.argument<Int>("timeoutMs") ?: 30000
                    server.scanFeature.startScan(timeout)
                        .fold({ result.success(true) },
                              { result.error("SCAN_FAILED", it.message, null) })
                }

                "stopScan" -> { server.scanFeature.stopScan(); result.success(true) }

                "isScanning" -> result.success(server.scanFeature.isScanning())

                "connect" -> {
                    val mac = call.argument<String>("address")
                        ?: return result.error("BAD_ARG", "address required", null)
                    server.connectFeature.connect(
                        bleAddress = mac,
                        edrAddress = call.argument<String>("edrAddr"),
                        deviceType = call.argument<Int>("deviceType") ?: -1,
                        connectWay = call.argument<Int>("connectWay") ?: 0,
                    ).fold({ result.success(true) },
                           { result.error("CONNECT_FAILED", it.message, null) })
                }

                "disconnect" -> {
                    val mac = call.argument<String>("address")
                        ?: return result.error("BAD_ARG", "address required", null)
                    server.connectFeature.disconnect(mac)
                        .fold({ result.success(true) },
                              { result.error("DISCONNECT_FAILED", it.message, null) })
                }

                "isConnected" -> {
                    val mac = call.argument<String>("address")
                        ?: return result.error("BAD_ARG", "address required", null)
                    result.success(server.connectFeature.isConnected(mac))
                }

                "connectedDevice" -> {
                    val d = server.connectFeature.connectedDevice()
                    if (d == null) result.success(null)
                    else result.success(
                        mapOf(
                            "address" to d.address,
                            "name" to runCatching { d.name }.getOrNull()
                        )
                    )
                }

                "deviceSnapshot" -> {
                    val mac = call.argument<String>("address")
                        ?: return result.error("BAD_ARG", "address required", null)
                    result.success(server.deviceInfoFeature.snapshot(mac))
                }

                "queryTargetInfo" -> {
                    val mac = call.argument<String>("address")
                        ?: return result.error("BAD_ARG", "address required", null)
                    val mask = call.argument<Int>("mask") ?: 0x0F
                    server.deviceInfoFeature.queryTargetInfo(
                        mac, mask,
                        onSuccess = { result.success(it) },
                        onError = { c, m -> result.error("TARGET_INFO_ERR", m, c) }
                    )
                }

                "sendCustomCmd" -> {
                    val mac = call.argument<String>("address")
                        ?: return result.error("BAD_ARG", "address required", null)
                    val opCode = call.argument<Int>("opCode")
                        ?: return result.error("BAD_ARG", "opCode required", null)
                    val payload = call.argument<ByteArray>("payload") ?: ByteArray(0)
                    server.customCmdFeature.send(
                        mac, opCode, payload,
                        onSuccess = { result.success(it) },
                        onError = { c, m -> result.error("CUSTOM_CMD_ERR", m, c) }
                    )
                }

                "startTranslation" -> {
                    val modeId = call.argument<Int>("modeId")
                        ?: return result.error("BAD_ARG", "modeId required", null)
                    @Suppress("UNCHECKED_CAST")
                    val args = (call.argument<Map<String, Any?>>("args") ?: emptyMap())
                    server.translationFeature.start(modeId, args)
                        .fold({ result.success(true) },
                              { result.error("TRANSLATION_ERR", it.message, null) })
                }

                "stopTranslation" -> {
                    server.translationFeature.stop(); result.success(true)
                }

                "translationStatus" -> result.success(
                    mapOf(
                        "working" to server.translationFeature.isWorking(),
                        "modeId" to server.translationFeature.currentModeId(),
                        "inputStreams" to server.translationFeature.currentInputStreams(),
                        "outputStreams" to server.translationFeature.currentOutputStreams()
                    )
                )

                "feedTranslatedAudio" -> {
                    val streamId = call.argument<String>("streamId")
                        ?: return result.error("BAD_ARG", "streamId required", null)
                    val pcm = call.argument<ByteArray>("pcm")
                        ?: return result.error("BAD_ARG", "pcm required", null)
                    val sampleRate = call.argument<Int>("sampleRate") ?: 16000
                    val channels = call.argument<Int>("channels") ?: 1
                    val bits = call.argument<Int>("bitsPerSample") ?: 16
                    val isFinal = call.argument<Boolean>("final") ?: false
                    val ok = server.translationFeature.feedTranslatedAudio(
                        streamId, pcm,
                        com.jielihome.jielihome.feature.translation.AudioFormat(sampleRate, channels, bits),
                        isFinal
                    )
                    result.success(ok)
                }

                "feedTranslationResult" -> {
                    server.translationFeature.feedTranslationResult(
                        srcLang = call.argument<String>("srcLang"),
                        srcText = call.argument<String>("srcText"),
                        destLang = call.argument<String>("destLang"),
                        destText = call.argument<String>("destText"),
                        requestId = call.argument<String>("requestId"),
                    )
                    result.success(true)
                }

                "isSupportCallTranslationWithStereo" -> {
                    val mac = call.argument<String>("address")
                    result.success(server.translationFeature.isSupportCallTranslationWithStereo(mac))
                }

                "feedAudioFilePcm" -> {
                    val pcm = call.argument<ByteArray>("pcm")
                        ?: return result.error("BAD_ARG", "pcm required", null)
                    val sampleRate = call.argument<Int>("sampleRate") ?: 16000
                    result.success(server.translationFeature.feedAudioFilePcm(pcm, sampleRate))
                }

                "speechIsRecording" ->
                    result.success(server.speechFeature.isRecording(call.argument<String>("address")))

                "speechStart" -> {
                    server.speechFeature.start(
                        address = call.argument<String>("address"),
                        voiceType = call.argument<Int>("voiceType")
                            ?: com.jieli.bluetooth.bean.record.RecordParam.VOICE_TYPE_OPUS,
                        sampleRate = call.argument<Int>("sampleRate")
                            ?: com.jieli.bluetooth.bean.record.RecordParam.SAMPLE_RATE_16K,
                        vadWay = call.argument<Int>("vadWay")
                            ?: com.jieli.bluetooth.bean.record.RecordParam.VAD_WAY_DEVICE,
                        onResult = { ok, msg ->
                            if (ok) result.success(true)
                            else result.error("SPEECH_START_ERR", msg, null)
                        }
                    )
                }

                "speechStop" -> {
                    server.speechFeature.stop(
                        address = call.argument<String>("address"),
                        reason = call.argument<Int>("reason")
                            ?: com.jieli.bluetooth.bean.record.RecordState.REASON_NORMAL,
                        onResult = { ok, msg ->
                            if (ok) result.success(true)
                            else result.error("SPEECH_STOP_ERR", msg, null)
                        }
                    )
                }

                "otaStart" -> {
                    val path = call.argument<String>("firmwareFilePath")
                        ?: return result.error("BAD_ARG", "firmwareFilePath required", null)
                    server.otaFeature.start(
                        address = call.argument<String>("address"),
                        firmwareFilePath = path,
                        blockSize = call.argument<Int>("blockSize") ?: 512,
                        fileFlagBytes = call.argument<ByteArray>("fileFlag") ?: ByteArray(0),
                    )
                    result.success(true)
                }

                "otaCancel" -> { server.otaFeature.cancel(); result.success(true) }

                "otaIsRunning" -> result.success(server.otaFeature.isRunning())

                else -> result.notImplemented()
            }
        } catch (t: Throwable) {
            result.error("PLUGIN_ERR", t.message, t.stackTraceToString())
        }
    }
}

File: app/local_plugins/devices/device_jieli/ios/device_jieli/Sources/device_jieli/MethodRouter.swift
import Flutter
import Foundation

/// 与 Android 端 `MethodRouter` 同语义：把 Dart MethodChannel 调用映射到 server 内的各 feature 模块。
/// 任何 throws/exception 统一兜底 `PLUGIN_ERR`，避免 Dart 端拿到悬空 Future。
public final class MethodRouter {

    private let server: JieliHomeServer
    public init(server: JieliHomeServer) { self.server = server }

    public func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = (call.arguments as? [String: Any]) ?? [:]
        do {
            try route(method: call.method, args: args, result: result)
        } catch {
            result(FlutterError(code: "PLUGIN_ERR", message: error.localizedDescription, details: nil))
        }
    }

    private func route(method: String, args: [String: Any], result: @escaping FlutterResult) throws {
        switch method {
        case "getPlatformVersion":
            result("iOS \(UIDevice.current.systemVersion)")

        case "initialize":
            server.initialize(
                multiDevice: args["multiDevice"] as? Bool ?? true,
                skipNoNameDev: args["skipNoNameDev"] as? Bool ?? false,
                enableLog: args["enableLog"] as? Bool ?? false
            )
            result(true)

        // ───── Scan ─────
        case "startScan":
            let timeoutMs = (args["timeoutMs"] as? NSNumber)?.intValue ?? 30_000
            let nameList = args["nameList"] as? [String] ?? []
            let uuidList = args["uuidList"] as? [String] ?? []
            let skipUnnamed = args["skipUnnamed"] as? Bool ?? true
            do {
                try server.scanFeature.startScan(
                    timeoutMs: timeoutMs,
                    nameList: nameList,
                    uuidList: uuidList,
                    skipUnnamed: skipUnnamed
                )
                result(true)
            } catch {
                result(FlutterError(code: "SCAN_FAILED", message: error.localizedDescription, details: nil))
            }

        case "stopScan":
            server.scanFeature.stopScan()
            result(true)

        case "isScanning":
            result(server.scanFeature.isScanning)

        // ───── Connect ─────
        case "connect":
            guard let address = args["address"] as? String, !address.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "address required", details: nil)); return
            }
            server.connectFeature.connect(
                bleAddress: address,
                edrAddress: args["edrAddr"] as? String,
                deviceType: (args["deviceType"] as? NSNumber)?.intValue ?? -1,
                connectWay: (args["connectWay"] as? NSNumber)?.intValue ?? 0,
                // Dart 侧明确传 dualConnect=true 才做 BR/EDR 桥接升级，默认纯 BLE。
                dualConnect: (args["dualConnect"] as? Bool) ?? false
            ) { ok, errMsg in
                if ok { result(true) }
                else { result(FlutterError(code: "CONNECT_FAILED", message: errMsg ?? "connect failed", details: nil)) }
            }

        case "disconnect":
            guard let address = args["address"] as? String, !address.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "address required", details: nil)); return
            }
            server.connectFeature.disconnect(address: address) { ok, errMsg in
                if ok { result(true) }
                else { result(FlutterError(code: "DISCONNECT_FAILED", message: errMsg ?? "disconnect failed", details: nil)) }
            }

        case "isConnected":
            guard let address = args["address"] as? String, !address.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "address required", details: nil)); return
            }
            result(server.connectFeature.isConnected(address: address))

        case "connectedDevice":
            if let info = server.connectFeature.connectedDeviceInfo() {
                result(info)
            } else {
                result(nil)
            }

        // ───── Device info ─────
        case "deviceSnapshot":
            guard let address = args["address"] as? String, !address.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "address required", details: nil)); return
            }
            result(server.deviceInfoFeature.snapshot(address: address))

        case "queryTargetInfo":
            guard let address = args["address"] as? String, !address.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "address required", details: nil)); return
            }
            let mask = (args["mask"] as? NSNumber)?.intValue ?? 0x0F
            server.deviceInfoFeature.queryTargetInfo(address: address, mask: mask) { info, errCode, errMsg in
                if let info = info { result(info) }
                else { result(FlutterError(code: "TARGET_INFO_ERR", message: errMsg ?? "query failed", details: errCode)) }
            }

        // ───── Custom cmd ─────
        case "sendCustomCmd":
            guard let address = args["address"] as? String, !address.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "address required", details: nil)); return
            }
            guard let opCode = (args["opCode"] as? NSNumber)?.intValue else {
                result(FlutterError(code: "BAD_ARG", message: "opCode required", details: nil)); return
            }
            let payloadBytes: [UInt8] = {
                if let arr = args["payload"] as? [Int] { return arr.map { UInt8(truncatingIfNeeded: $0) } }
                if let arr = args["payload"] as? [NSNumber] { return arr.map { UInt8(truncatingIfNeeded: $0.intValue) } }
                if let typed = args["payload"] as? FlutterStandardTypedData {
                    return [UInt8](typed.data)
                }
                return []
            }()
            server.customCmdFeature.send(address: address, opCode: opCode, payload: Data(payloadBytes)) { ok, data, errCode, errMsg in
                if ok {
                    result(data.map { [UInt8]($0) } ?? [])
                } else {
                    result(FlutterError(code: "CUSTOM_CMD_ERR", message: errMsg ?? "custom cmd failed", details: errCode))
                }
            }

        // ───── Translation ─────
        case "startTranslation":
            guard let modeId = (args["modeId"] as? NSNumber)?.intValue else {
                result(FlutterError(code: "BAD_ARG", message: "modeId required", details: nil)); return
            }
            let innerArgs = args["args"] as? [String: Any] ?? [:]
            print("[JieliRouter] startTranslation modeId=\(modeId) args=\(innerArgs) " +
                  "assistantRunning=\(server.assistantBridge.isRunning) " +
                  "deviceRecording=\(server.deviceRecordFeature.isRecording) " +
                  "translationWorking=\(server.translationFeature.isWorking)")
            // 设备录音 / AI 助理 与翻译互斥（同一 JLTranslationManager session）
            if server.assistantBridge.isRunning {
                print("[JieliRouter] startTranslation: stopping AssistantBridge first")
                server.assistantBridge.stop()
            }
            if server.deviceRecordFeature.isRecording {
                print("[JieliRouter] startTranslation: stopping DeviceRecord first")
                server.deviceRecordFeature.stop()
            }
            server.translationFeature.start(modeId: modeId, args: innerArgs) { ok, errMsg in
                print("[JieliRouter] startTranslation result modeId=\(modeId) ok=\(ok) err=\(errMsg ?? "nil")")
                if ok { result(true) }
                else { result(FlutterError(code: "TRANSLATION_ERR", message: errMsg ?? "start failed", details: nil)) }
            }

        case "stopTranslation":
            server.translationFeature.stop()
            result(true)

        case "translationStatus":
            result([
                "working": server.translationFeature.isWorking,
                "modeId": server.translationFeature.currentModeId as Any,
                "inputStreams": server.translationFeature.currentInputStreams,
                "outputStreams": server.translationFeature.currentOutputStreams,
            ] as [String: Any?])

        case "feedTranslatedAudio":
            guard let streamId = args["streamId"] as? String else {
                result(FlutterError(code: "BAD_ARG", message: "streamId required", details: nil)); return
            }
            guard let pcm = (args["pcm"] as? FlutterStandardTypedData)?.data else {
                result(FlutterError(code: "BAD_ARG", message: "pcm required", details: nil)); return
            }
            let sr = (args["sampleRate"] as? NSNumber)?.intValue ?? 16000
            let ch = (args["channels"] as? NSNumber)?.intValue ?? 1
            let bits = (args["bitsPerSample"] as? NSNumber)?.intValue ?? 16
            let isFinal = args["final"] as? Bool ?? false
            // 诊断：只在段尾或路由失败时打。
            let ok = server.translationFeature.feedTranslatedAudio(
                streamId: streamId,
                pcm: pcm,
                sampleRate: sr, channels: ch, bitsPerSample: bits,
                isFinal: isFinal
            )
            if isFinal || !ok {
                NSLog("[MethodRouter] feedTranslatedAudio streamId=%@ pcm=%dB isFinal=%@ routed=%@",
                      streamId, pcm.count, isFinal ? "true" : "false", ok ? "true" : "false")
            }
            result(ok)

        case "feedTranslationResult":
            server.translationFeature.feedTranslationResult(
                srcLang: args["srcLang"] as? String,
                srcText: args["srcText"] as? String,
                destLang: args["destLang"] as? String,
                destText: args["destText"] as? String,
                requestId: args["requestId"] as? String
            )
            result(true)

        case "isSupportCallTranslationWithStereo":
            result(server.translationFeature.isSupportCallTranslationWithStereo(address: args["address"] as? String))

        case "feedAudioFilePcm":
            guard let pcm = (args["pcm"] as? FlutterStandardTypedData)?.data else {
                result(FlutterError(code: "BAD_ARG", message: "pcm required", details: nil)); return
            }
            let sr = (args["sampleRate"] as? NSNumber)?.intValue ?? 16000
            result(server.translationFeature.feedAudioFilePcm(pcm: pcm, sampleRate: sr))

        // ───── Speech assistant (cmd=4/5/210 路径) ─────
        case "speechIsRecording":
            result(server.speechFeature.isRecording(address: args["address"] as? String))

        case "speechStart":
            server.speechFeature.start(
                address: args["address"] as? String,
                voiceType: (args["voiceType"] as? NSNumber)?.intValue ?? 2,
                sampleRate: (args["sampleRate"] as? NSNumber)?.intValue ?? 16,
                vadWay: (args["vadWay"] as? NSNumber)?.intValue ?? 0
            ) { ok, msg in
                if ok { result(true) }
                else { result(FlutterError(code: "SPEECH_START_ERR", message: msg ?? "speech start failed", details: nil)) }
            }

        case "speechStop":
            server.speechFeature.stop(
                address: args["address"] as? String,
                reason: (args["reason"] as? NSNumber)?.intValue ?? 0
            ) { ok, msg in
                if ok { result(true) }
                else { result(FlutterError(code: "SPEECH_STOP_ERR", message: msg ?? "speech stop failed", details: nil)) }
            }

        // ───── AI 助理通路 ─────
        case "assistantStart":
            if server.translationFeature.isWorking { server.translationFeature.stop() }
            if server.deviceRecordFeature.isRecording { server.deviceRecordFeature.stop() }
            let ok = server.assistantBridge.start(
                address: args["address"] as? String,
                sampleRate: (args["sampleRate"] as? NSNumber)?.intValue ?? 16000
            )
            if ok { result(true) }
            else { result(FlutterError(code: "ASSISTANT_START_ERR", message: "assistant bridge start failed", details: nil)) }

        case "assistantStop":
            server.assistantBridge.stop()
            result(true)

        case "assistantIsRunning":
            result(server.assistantBridge.isRunning)

        // ───── 设备录音 ─────
        case "deviceRecordStart":
            let innerArgs = args["args"] as? [String: Any] ?? [:]
            if server.translationFeature.isWorking { server.translationFeature.stop() }
            if server.assistantBridge.isRunning { server.assistantBridge.stop() }
            server.deviceRecordFeature.start(args: innerArgs) { ok, errMsg in
                if ok { result(true) }
                else { result(FlutterError(code: "DEVICE_RECORD_ERR", message: errMsg ?? "device record start failed", details: nil)) }
            }

        case "deviceRecordStop":
            server.deviceRecordFeature.stop()
            result(true)

        case "deviceRecordStatus":
            result(["recording": server.deviceRecordFeature.isRecording])

        // ───── OTA ─────
        case "otaStart":
            guard let path = args["firmwareFilePath"] as? String, !path.isEmpty else {
                result(FlutterError(code: "BAD_ARG", message: "firmwareFilePath required", details: nil)); return
            }
            server.otaFeature.start(
                address: args["address"] as? String,
                firmwareFilePath: path,
                blockSize: (args["blockSize"] as? NSNumber)?.intValue ?? 512,
                fileFlagBytes: (args["fileFlag"] as? FlutterStandardTypedData)?.data ?? Data()
            )
            result(true)

        case "otaCancel":
            server.otaFeature.cancel()
            result(true)

        case "otaIsRunning":
            result(server.otaFeature.isRunning)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
